File: Example/TTPatchUnitTests.js
/**
 * 引入UI组件,不引入无法直接使用
 */
_import('TTPatchURLSession,NSURLRequest,NSURL,NSString,TTPatchHotRefrshTool,UIDevice,UIView,UILabel,UIColor,UIFont,UIScreen,UIImageView,UIImage,UITapGestureRecognizer,UIButton,TTPlaygroundModel')

/**
 *  @params:1.要替换的Class名,`:`标识继承关系
 *  @params:2.声明实例方法
 *  @params:3.声明静态方法
 *  声明Class,如无需在Oc中动态创建,可不设置父类,直接在js中创建类
 *  声明Class,如Native不存在,则动态创建Class
 */
defineClass('TTPatchUnitTests:XCTestCase', {
    /**
     * 添加属性,自动生成`setter`/`getter`方法,取值和赋值必须使用`setter`/`getter`方法.
     */
    name: property(),

    blockAddSignatureCase: function () {
        self.testInt_string_obj_(1,"TTPatch");
        var list = [1,2,3,];
        var dic = {"key":"val"};
        // self.testArray_dic_(list,dic);

        // self.testBlockInt_(block("void,int",function(){

        // }))
        // testInt_string_obj_: function(p1, p2, p3) {},
        // testArray_dic_: function(p1, p2) {},
        // testBlockInt_: function(p1) {},
        // testBlockString_: function(p1) {},
        // testBlockObj_: function(p1) {},
            self.testCall0_(block("", function () {
                Utils.log_info('[unit-test] --------JS传入OC方法,接受到回调--------- 无参数,无返回值');
            }));
            self.testCall1_(block('void,id,int', function (arg1, arg2) {
                var dic = JSON.parse(arg1);
                Utils.log_info('[unit-test] --------JS传入OC方法,接受到回调--------- 有参数,无返回值  ' + dic.name + '-' + arg2);
            }));


            self.testCall2_(block("id,id", function (arg) {
                Utils.log_info('[unit-test] --------JS传入OC方法,接受到回调--------- 有参数,有返回值:string  ' + arg);
                // arg.view().setBackgroundColor_(UIColor.blackColor());
                return arg;
            }));


            self.testCall3_(block("id,void", function () {
                Utils.log_info('[unit-test] --------JS传入OC方法,接受到回调--------- 无参数,有返回值:string  ');
                return '这是有返回值的哦';
            }));
        },
        funcWithBlockParams_param2_: dynamic('void,id,?', function (arg1, callback) {
            Utils.log_info('[unit-test] [1]funcWithBlockParams:' + arg1);
            if (callback) {
                callback("1");
            }
        }),
    funcWithBlockParams_paramInt2_: dynamic('void,id,?', function (arg1, callback) {
        Utils.log_info('[unit-test] [2]funcWithBlockParams:paramInt2:' + arg1);
        if (callback) {
            callback(2);
        }
    }),
    funcWithParams_param2_: dynamic('void,id,id', function (param1, param2) {
        var arg1 = param1[0];
        var arg2 = param2['key'];
        Utils.log_info('[unit-test] [3]funcWithParams_param2_:' + param1 + ',' + arg2);
    }),
    testBlockInt_: dynamic('void,int',function(p1) {
        p1(1);
        Utils.log_info('[unit-test] [3]testBlockInt_:' + p1);
    }),
    testBlockString_: dynamic('void,@',function(p1) {
        p1("ttpatch");
        Utils.log_info('[unit-test] [3]testBlockString_:' + p1);
    }),
    testBlockObj_: dynamic('void,@',function(p1) {
        var dic = {"key":"val"};
        p1(dic);
        Utils.log_info('[unit-test] [3]testBlockObj_:' + p1);
    })
}, {
    //静态方法
    testAction_: function (str) {Utils.log_info('[unit-test] 静态方法调用成功-> ' + str);}
})

defineClass('UserModel:NSObject', {
    getUserName: function () {
            Utils.log_info(self.getUserPW());
            return "Alibaba";
        },
        getUserPW: function () {
            return "self -> UserModel";
        },
}, {});

File: Example/TTPatchUnitTests.m
//
//  TTPatchUnitTests.m
//  ExampleTests
//
//  Created by tianyubing on 2020/4/24.
//  Copyright © 2020 TianyuBing. All rights reserved.
//

#import "TTPatchUnitTests.h"
#import "TTPatch.h"
#import "TTPatchKit.h"
#import "TTPatchHotRefrshTool.h"

@implementation TTPatchUnitTests

- (void)setUp {
    // Put setup code here. This method is called before the invocation of each test method in the class.
    // 初始化SDK
//    [TTPatch initSDK];

}

- (void)tearDown {
    // Put teardown code here. This method is called after the invocation of each test method in the class.
}

- (void)testExample {
    // This is an example of a functional test case.
    // Use XCTAssert and related functions to verify your tests produce the correct results.

        /**
//         * 连接本地测试服务,如加载空白,请检查
//         * 1.本地服务是否已启动成功
//         * 2.检查`info.plist`中IP是否获取正确
//         */
//        [self testSocket];
//        // 拉取本地js资源
////        [self updateResource:@"hotfixPatch" callbacl:nil];
////        [self loadupdateResource:@"hotfixPatch.js" callbacl:nil];
////    [TTPatchUnitTests testAction:@"unit-test"];
    [self testNativeCallJsCase];
}


- (void)testNativeCallJsCase{
//    sleep(10);
    /*
     * native调用js方法,参数带block
     * 实际场景:
     *      可将线上出bug的方法替换为js实现,从而规避bug.
     */
    [self testBlockInt:^(int p1) {
        TTLog_Info(@"[%s] p1:%d",__func__,p1);
    }];
    
    [self testBlockString:^(NSString * _Nonnull p1) {
        TTLog_Info(@"[%s] p1:%@",__func__,p1);
    }];
    
    [self testBlockObj:^(id _Nonnull p1) {
       TTLog_Info(@"[%s] p1:%@",__func__,p1);
    }];
}

/*
 * js调用native方法,参数携带block
 * 实际场景:
 *      js可以实现一部分逻辑,如必须调用native方法,可以参照该场景实现
 */
- (void)testCall0:(void(^)(void))call{
    if (call) {
        call();
    }
}

- (void)testInt:(int)p1 string:(NSString*)p2{
    
}
- (void)testArray:(NSArray *)p1 dic:(NSDictionary *)p2{
    
}

//- (void)testBlockInt:(void(^)(int))p1{
//
//}
//- (void)testBlockString:(void(^)(NSString*))p1{
//
//}
//- (void)testBlockObj:(void(^)(id))p1{
//
//}

- (void)testCall1:(void(^)(NSString * str,int inta))call{
    if (call) {
        call(@"{\"id\":1,\"name\":\"Tencent\",\"email\":\"admin@Tencent.com\",\"interest\":[\"Tencent\",\"Tencent\"]}",999);
    }
}

- (void)testCall2:(NSString *(^)(NSString *str))call{
    if (call) {
        NSString * val = call(@"input");
        
    }
}

- (void)testCall3:(NSString *(^)(void))call{
    if (call) {
        NSString * val = call();
        
    }
}

/*
* js调用native方法,静态方法
*/
+ (void)testAction:(NSString *)str{
    
}


- (void)testSocket{
    
    NSString *socket;
#if TARGET_IPHONE_SIMULATOR  //模拟器
    socket = [NSString stringWithFormat:@"ws://%@:%@/socket.io/?EIO=4&transport=websocket",
              @"127.0.0.1",
              [TTPatchHotRefrshTool shareInstance].getLocaServerPort];
#elif TARGET_OS_IPHONE      //真机
    socket = [NSString stringWithFormat:@"ws://%@:%@/socket.io/?EIO=4&transport=websocket",
              [TTPatchHotRefrshTool shareInstance].getLocaServerIP,
              [TTPatchHotRefrshTool shareInstance].getLocaServerPort];
#endif
    [[TTPatchHotRefrshTool shareInstance] startLocalServer:socket];
    [TTPatchHotRefrshTool shareInstance].delegate = self;
}

- (void)reviceRefresh:(id)msg{
//    [self updateResource:msg callbacl:nil];
}

//- (void)updateResource:(NSString *)filename callbacl:(void(^)(void))callback
//{
//    NSString *srcPath = [[NSBundle mainBundle] pathForResource:filename ofType:@"js"];
//
//
//    NSString *jsCode = [[NSString alloc] initWithData:[[NSFileManager defaultManager] contentsAtPath:srcPath] encoding:NSUTF8StringEncoding];
//
//    [[TTPatch shareInstance] evaluateScript:[[TTPatch shareInstance] formatterJS:jsCode] withSourceURL:[NSURL URLWithString:filename]];
//}


@end
